design/multicheckout.md

```diff
@@ -0,0 +1,118 @@
+# Multiple repo checkout
+
+## Scenarios
+
+1. My code is in one repository, but I have tools / scripts / dependencies in another repo that I need to build my code.
+2. My code is one repository, but my pipeline YAML is in another.
+3. My code is split up across multiple repos.
+4. I want to control the checkout location of my code.
+
+## YAML
+
+```yaml
+resources:
+  repositories:
+  - repository: tools
+    type: git
+    name: ToolsRepo
+  - repository: scripts
+    type: github
+    name: ContosoOrg/ScriptsRepo
+    endpoint: GitHubApp1
+
+steps:
+- checkout: self
+- checkout: tools
+  root: buildTools
+- checkout: scripts
+```
+
+In the above example, three repos will end up checked out.
+`self` is the usual repo where the pipeline YAML was found.
+It will be checked out in the default location.
+`tools` is another repo in the same Azure DevOps project.
+It will be checked out into `buildTools/`, a subdirectory of a new "resources" directory.
+Finally, `scripts` is a repo on GitHub.
+It will be checked out into its default location, `scripts/` (driven by the resource name) underneath the resources directory.
+
+## Schema
+
+```yaml
+- checkout: none | self | resourceName
+  clean: boolean      # whether to fetch clean each time
+  fetchDepth: number  # the depth of commits to ask Git to fetch
+  lfs: boolean        # whether to download Git-LFS files
+  submodules: true | recursive  # set to 'true' for a single level of submodules or 'recursive' to get submodules of submodules
+  persistCredentials: boolean   # set to 'true' to leave the OAuth token in the Git config after the initial fetch
+  root: string        # directory to checkout the repo
+```
+
+## New features needed
+
+- Resources directory: $(Pipelines.ResourcesDirectory)
+- Default source directory: $(Pipelines.DefaultSourcesDirectory)
+- `self` checkout influences $(Pipelines.SourcesDirectory)
```

Comment:
why `self` influences `$(Pipelines.SourcesDirectory)`? it would make more sense to me it influences `$(Pipelines.DefaultSourcesDirectory)`

Reply:
what's the different between these 2 folders.
here is the the agent work folder layout looks like:
```
D:\_work\1\s
                 \s\self
                 \s\toolRepo
                 \s\scriptRepo
```
which folder is $(Pipelines.ResourcesDirectory) and which folder is $(Pipelines.DefaultSourcesDirectory)?

Reply:
ResourcesDirectory would be a new concept. Maybe `d:\_work\1\r`. By default, self goes in `d:\_work\1\s` and a repo called `foo` goes in `d:\_work\1\r\foo`.

DefaultSourcesDirectory is static - on Hosted, always `d:\_work\1\s`. SourcesDirectory is the actual place where your `self` repo was placed, so in your example, `d:\_work\1\s\self`. In a single-checkout case with no specified `path`, the two folders are identical.

If this design works, I'll clarify in the spec. If you're asking for changes, what do you propose?

Reply:
If repo self goes to d:\_work\1\s then why wouldn't foo go to d:\_work\1\foo?  I am not sure I would put repos in the resources folder as that will simply make things difficult to use for the primary scenario of needing multiple repos for a build.

Reply:
I chose the resources directory for non-`self` repo to match what other resources will do. What do we do if the user starts giving repo resources names like "a"? I figured we own the top dirs (_work\\1\\r, _work\\1\\s, etc.) and then the user has an empty namespace inside to work with.

Reply:
I updated the spec. Now, I propose we put everything under the current `\s` sources directory in its own directory by default.